lib/rate_limited_request_map.js

```diff
@@ -0,0 +1,31 @@
+'use strict';
+
+function RateLimitedRequestMap() {
+  this._requests = new Map();
+}
+
+RateLimitedRequestMap.prototype.keepWorse = function add(rlRequest) {
+  if (!rlRequest) { return; }
+
+  const requestId = rlRequest.id;
+  const old = this._requests.get(requestId);
+
+  if (old && rlRequest.hasMoreRemainingThan(old)) { return old; }
+
+  this._requests.set(requestId, rlRequest);
+
+  return rlRequest;
+};
+
+RateLimitedRequestMap.prototype.get = function get(requestId) {
+  return this._requests.get(requestId);
+};
+
+RateLimitedRequestMap.prototype.getAndRemove = function getAndRemove(requestId) {
+  const rlRequest = this.get(requestId);
```

Comment:
I'm not sure what you mean... this function is more or less like `pop` in arrays, I need to return the deleted element. Map#delete returns a boolean.

From doc:
> Returns true if an element in the Map object existed and has been removed, or false if the element does not exist.